### DISCUSSION TITLE
Support for reactified zoteropane

### DISCUSSION THREAD

**Discussion by retorquere** [retorquere/zotero-better-bibtex#1904] [General]
@dstillman, @adomasven, in `_renderCell`, must I always return a span with two subspans, with those classnames? or can I just return a single span with className `cell ${column.className} cell-text`?

**Comment by adomasven** [comment]
> @dstillman, @adomasven, in `_renderCell`, must I always return a span with two subspans, with those classnames? or can I just return a single span with className `cell ${column.className} cell-text`?

If you just want the text you only need to override `Item.getField()`. You override `_renderCell()` if you need to perform custom cell styling and rendering. You can look for rendering functions in zotero's `itemTree.jsx` and `virtualized-table.jsx` to get an idea for what's going on there.

> Which of the css changes in the scite PR are relevant to all plugins rather than specifically to scite?

If you're just doing text in your column you don't need any additional css. Otherwise add styling as you see fit to customize BBT's cells.

**Reply by retorquere** [reply]
In `_renderCell`, how can I pass eg data about whether the citekey is pinned?

**Reply by retorquere** [reply]
Do I still have to patch `_getRowData` if I have patched `getField`?

**Reply by adomasven** [reply]
> In `_renderCell`, how can I pass eg data about whether the citekey is pinned?

Could you elaborate? What column is this displayed in and what should the cell contain?

> Do I still have to patch `_getRowData` if I have patched `getField`?

No, in fact that might not work as I described in that original PR and you in fact need to patch `getField()` instead, or Zotero will throw in `_getRowData()`. I might need to slightly change the signature of the function for it to be possible to monkey patch.

**Reply by retorquere** [reply]
> > In `_renderCell`, how can I pass eg data about whether the citekey is pinned?
> 
> Could you elaborate? What column is this displayed in and what should the cell contain?

In `getCellText`, I currently append `\uD83D\uDCCC` to the citekey to show it is pinned, but in `getCellText` I have the item ID at hand so I can do this lookup. I don't now what `data` looks like in `_renderCell` but from the scite patch it looks like it's just text. So I don't know how to find out what item is being rendered.

> No, in fact that might not work as I described in that original PR and you in fact need to patch `getField()` instead, or Zotero will throw in `_getRowData()`. I might need to slightly change the signature of the function for it to be possible to monkey patch.

I was just about to ask about this ;) but then I still don't know how to display the pin icon in the items list. I don't want to return the pin icon in `getField`.

edit: this is probably why I'd want to override `_renderCell` then?

**Reply by adomasven** [reply]
> In `getCellText`, I currently append `\uD83D\uDCCC` to the citekey to show it is pinned, but in `getCellText` I have the item ID at hand so I can do this lookup. I don't now what `data` looks like in `_renderCell` but from the scite patch it looks like it's just text. So I don't know how to find out what item is being rendered.

You can use `this.getRow(index).ref` to get the item in `_renderCell()`. You could also return the pin in getField, but it's more appropriate in the render function.

**Reply by retorquere** [reply]
and `data` is then just the return value from the `getField` for the column being rendered?

**Reply by retorquere** [reply]
Doesn't look like it -- looks like it's empty for me? I can just ignore `data` and get the citekey from the `item.id` (I'm looking it up anyway) but then I don't know what `data` is for.

**Reply by adomasven** [reply]
`data` is supposed to be what you returned in `getField()`. It definitely works, so perhaps something goes wrong in your `getField()` override.

**Reply by retorquere** [reply]
`data` is indeed filled; the problem is that the I can't query my own database until after my init is done, so I have

```
if (Zotero.BetterBibTeX.ready.isPending()) return ''
```

in the getField. Once my init is done, `getField` will start returning the citekeys, but I don't currently know how to tell zotero "refresh every row". If I click a group library and go back to my own library, the keys are displayed.

**Reply by adomasven** [reply]
Call `ZoteroPane.itemsView.tree.invalidate()`

**Reply by retorquere** [reply]
That doesn't change anything for me, the behavior stays the same. When I edit an item in any way, the citekey is displayed, and if I click a different library and back to my library, all citekeys are displayed. Calling `ZoteroPane.itemsView.tree.invalidate()` doesn't yield an error, but also no refresh.

**Reply by retorquere** [reply]
After I call invalidate, I see no new calls to `getField`.

**Reply by adomasven** [reply]
Ok, sorry, that doesn't refresh the cached row info. The most appropriate thing to do would be to call `Zotero.Notifier.trigger('modify', 'item', [item ids]);` which is a generic way to tell Zotero to update its UI for modified items, which causes them to be repainted wherever they're displayed. Otherwise you can call `ZoteroPane.itemsView.refreshAndMaintainSelection()`.

**Reply by dstillman** [reply]
Don't send a 'modify' notification yourself — Zotero does that automatically when objects are actually changed in the DB. I haven't followed this whole thread, but a 'refresh' event might do what you want.

**Reply by retorquere** [reply]
I'm currently using the refreshAndMaintainSelection call, is that what you mean by refresh?

**Reply by dstillman** [reply]
No, `Zotero.Notifier.trigger('refresh', 'item', [item ids]);`

**Reply by adomasven** [reply]
Dan meant `Zotero.Notifier.trigger('refresh', 'item', [item ids])` which would be the most appropriate, but `refreshAndMaintainSelection()` is fine if you need to update the row data for every row.

**Reply by retorquere** [reply]
For that I'd have to collect the currently visible item ids. How would I do this?

**Reply by adomasven** [reply]
Since you need to refresh all visible rows it's probably more appropriate to just call `refreshAndMaintainSelection()`. The notifier is more intended for individual items. And to get the displayed rows you'd need to access private properties on the `itemsView`, which would the defeat the purpose of using "the more appropriate API call".

**Reply by retorquere** [reply]
Then the current build should do it. I'm awaiting feedback from users, but for me it works.

**Comment by retorquere** [comment]
I'm currently hitting

<details><summary>error log</summary>

```
JavaScript error: chrome://zotero/content/xpcom/data/itemFields.js, line 280: Error: Invalid field 'undefined' for base field
console.error: "The above error occurred in the <VirtualizedTable> component:\n    in VirtualizedTable (created by Context.Consumer)\n    in injectIntl(VirtualizedTable) (created by ForwardRef)\n    in ForwardRef (created by ItemTree)\n    in ItemTree\n    in IntlProvider\n\nReact will try to recreate this component tree from scratch using the error boundary you provided, ItemTree."
zotero(3)(+0000039): ItemTree: React threw an error

zotero(1)(+0000000): Error: An error was thrown inside one of your components, but React doesn't know what it was. This is likely due to browser flakiness. React does its best to preserve the "Pause on exceptions" behavior of the DevTools, which requires some DEV-mode only tricks. It's possible that these don't work in your browser. Try triggering the error in production mode, or switching to a modern browser. If you suspect that this is actually an issue with React, please file an issue.

    invokeGuardedCallbackDev@resource://zotero/loader.jsm -> resource://zotero/react-dom.js:240:21
    invokeGuardedCallback@resource://zotero/loader.jsm -> resource://zotero/react-dom.js:286:5
    commitRootImpl@resource://zotero/loader.jsm -> resource://zotero/react-dom.js:22676:11
    unstable_runWithPriority@resource://zotero/loader.jsm -> resource://zotero/react.js:2685:14
    runWithPriority$1@resource://zotero/loader.jsm -> resource://zotero/react-dom.js:11174:12
    commitRoot@resource://zotero/loader.jsm -> resource://zotero/react-dom.js:22516:5
    finishSyncRender@resource://zotero/loader.jsm -> resource://zotero/react-dom.js:21942:5
    performSyncWorkOnRoot@resource://zotero/loader.jsm -> resource://zotero/react-dom.js:21928:9
    flushSyncCallbackQueueImpl/<@resource://zotero/loader.jsm -> resource://zotero/react-dom.js:11224:26
    unstable_runWithPriority@resource://zotero/loader.jsm -> resource://zotero/react.js:2685:14
    runWithPriority$1@resource://zotero/loader.jsm -> resource://zotero/react-dom.js:11174:12
    flushSyncCallbackQueueImpl@resource://zotero/loader.jsm -> resource://zotero/react-dom.js:11219:9
    flushSyncCallbackQueue@resource://zotero/loader.jsm -> resource://zotero/react-dom.js:11207:5
    scheduleUpdateOnFiber@resource://zotero/loader.jsm -> resource://zotero/react-dom.js:21334:11
    enqueueForceUpdate@resource://zotero/loader.jsm -> resource://zotero/react-dom.js:12813:7
    Component.prototype.forceUpdate@resource://zotero/loader.jsm -> resource://zotero/react.js:576:5
    clearItemsPaneMessage/<@chrome://zotero/content/itemTree.js:3497:576
    clearItemsPaneMessage@chrome://zotero/content/itemTree.js:3497:553
    changeCollectionTreeRow@chrome://zotero/content/itemTree.js:3583:198
```

</details>


The changes for my react-compatibe branch sit [here](https://github.com/retorquere/zotero-better-bibtex/blob/react/content/better-bibtex.ts) BTW.

**Reply by adomasven** [reply]
I've added a comment here for what the issue there is. https://github.com/retorquere/zotero-better-bibtex/commit/04df6e665d5b3318d8c40cf39391f0a42000d2cc#r55605726

**Comment by retorquere** [comment]
In 

```
 icon.className = 'icon icon-bg cell-icon'
```

what stylesheet are these picking icons from?

**Reply by adomasven** [reply]
> In
> 
> ```
>  icon.className = 'icon icon-bg cell-icon'
> ```
> 
> what stylesheet are these picking icons from?

These are default classes from Zotero, we use those to style our own icons (item type, tags, etc). See https://github.com/zotero/zotero/blob/cbbff600a60c9e7a7407d6f2e4053309bf28b872/scss/components/_icons.scss
Scite icons are set in their .css file, and Zotero sets them inline in https://github.com/zotero/zotero/blob/67d9e87c55dfbbb4ced8c4dbe25c2a6ea09f44f0/chrome/content/zotero/components/icons.jsx

**Comment by retorquere** [comment]
I think I have now covered all changes mentioned in https://github.com/scitedotai/scite-zotero-plugin/pull/26. @dstillman, @adomasven , super thanks for navigating me through this. It's a great weight off my shoulder that this went so smoothly.

**Reply by adomasven** [reply]
Great! Thanks for taking the time to do this, a lot of Zotero users rely on BBT and this was very short notice.

**Reply by retorquere** [reply]
Is the `ZoteroPane` global new to the beta, or also available in the non-beta release? I've been using `Zotero.getActiveZoteroPane()` but I wondered for a long time whether there could be more than one in the post-firefox world.

**Comment by dstillman** [comment]
Not new — that's been around forever.

And yes, there could be more than one, but since that's not officially possible at the moment (aside from a weird hack on macOS), it's not something we've spent much time thinking about.

**Reply by dstillman** [reply]
The main difference is just that `ZoteroPane` exists on the `window`, but `Zotero.getActiveZoteroPane()` is available to XPCOM code.
